horovod/mxnet/handle_manager.cc

```diff
@@ -43,8 +43,16 @@ bool HandleManager::PollHandle(int handle) {
 }
 
 void HandleManager::ExecuteCallback(int handle) {
-  Callback cb = *callbacks_[handle];
-  cb();
+  std::unique_lock<std::mutex> lock(mutex_);
```

Comment:
Maybe using a lock_guard to lock and unlock this mutex.

Reply:
See Line 51 for unlocking. Once we retrieved the callback object, we can release it earlier.